docs/ql-design-patterns.md

```diff
@@ -0,0 +1,82 @@
+# CodeQL Design Patterns
+
+A list of design patterns you are recommended to follow.
+
+## `::Range` for extensibility and refinement
+
+To allow both extensibility and refinement of classes, we use what is commonly referred to as the `::Range` pattern (since https://github.com/github/codeql/pull/727), but the actual implementation can use different names.
+
+<details>
+<summary>Generic example of how to define classes with ::Range</summary>
+
+Instead of
+```ql
+/** <QLDoc...> */
+abstract class MySpecialExpr extends Expr {
+  /** <QLDoc...> */
+  abstract int memberPredicate();
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr { ... }
+```
+
+use
+
+```ql
+/**
+ * <QLDoc...>
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `MySpecialExpr::Range` instead.
+ */
+class MySpecialExpr extends Expr {
+  MySpecialExpr::Range range;
+
+  MySpecialExpr() { this = range }
+
+  /** <QLDoc...> */
+  int memberPredicate() { result = range.memberPredicate() }
+}
+
+/** Provides a class for modeling new <...> APIs. */
+module MySpecialExpr {
+  /**
+   * <QLDoc...>
+   *
+   * Extend this class to model new APIs. If you want to refine existing API models,
+   * extend `MySpecialExpr` instead.
+   */
+  abstract class Range extends Expr {
+    /** <QLDoc...> */
+    abstract int memberPredicate();
+  }
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr::Range { ... }
+```
```

Comment:
```suggestion
Now, a concrete subclass can derive from `MySpecialExpr::Range` if it wants to extend the set of values in `MySpecialExpr`, and it will be required to implement the abstract `memberPredicate()`. Conversely, if it wants to refine `MySpecialExpr` and override `memberPredicate` for all extensions, it can do so by deriving from `MySpecialExpr` directly.

The key element of the pattern is to provide a field of type `MySpecialExpr::Range`, equating it to `this` in the characteristic predicate of `MySpecialExpr`. In member predicates, we can use either `this` or `range`, depending on which type has the API we need.
```